crates/biome_js_analyze/src/lint/nursery/use_top_level_regex.rs

```diff
@@ -0,0 +1,80 @@
+use biome_analyze::{context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_js_syntax::{
+    JsInitializerClause, JsModuleItemList, JsRegexLiteralExpression, JsVariableDeclaration,
+    JsVariableDeclarator, JsVariableDeclaratorList, JsVariableStatement,
+};
+use biome_rowan::AstNode;
+
+declare_rule! {
+    /// Enforce all regex literals to be declared at the top level.
+    ///
+    /// This rule is useful to avoid performance issues when using regex literals inside functions. Regex literals are parsed every time they are used to create a new RegExp object. By declaring them at the top level, this overhead can be avoided.
+    ///
+    /// It's important to note that this rule is not recommended for all cases. Placing regex literals at the top level can hurt startup times. In browser contexts, this can result in longer page loads.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo(someString) {
+    ///     return /[a-Z]*/.test(someString)
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// const REGEX = /[a-Z]*/;
+    ///
+    /// function foo(someString) {
+    ///     return REGEX.test(someString)
+    /// }
+    /// ```
+    ///
+    pub UseTopLevelRegex {
+        version: "next",
+        name: "useTopLevelRegex",
+        recommended: false,
+    }
+}
+
+impl Rule for UseTopLevelRegex {
+    type Query = Ast<JsRegexLiteralExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let regex = ctx.query();
+        let parent = regex
+            .parent::<JsInitializerClause>()
+            .and_then(|node| node.parent::<JsVariableDeclarator>())
+            .and_then(|node| node.parent::<JsVariableDeclaratorList>())
+            .and_then(|node| node.parent::<JsVariableDeclaration>())
+            .and_then(|node| node.parent::<JsVariableStatement>())
+            .and_then(|node| node.parent::<JsModuleItemList>());
```

Comment:
We have a node called `AnyJsControlFlowRoot` that should list all the cases you need to cover. Also, I would use `syntax().ancestors()` instead of `parent()`